// @makerdao/dai/src/eth/EthereumCdpService.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _desc, _value, _class;

var _PrivateService2 = require('../core/PrivateService');

var _PrivateService3 = _interopRequireDefault(_PrivateService2);

var _contracts = require('../../contracts/contracts');

var _contracts2 = _interopRequireDefault(_contracts);

var _Cdp = require('./Cdp');

var _Cdp2 = _interopRequireDefault(_Cdp);

var _ProxyCdp = require('./ProxyCdp');

var _ProxyCdp2 = _interopRequireDefault(_ProxyCdp);

var _bignumber = require('bignumber.js');

var _bignumber2 = _interopRequireDefault(_bignumber);

var _constants = require('../utils/constants');

var _Currency = require('./Currency');

var _conversion = require('../utils/conversion');

var _tracksTransactions = require('../utils/tracksTransactions');

var _tracksTransactions2 = _interopRequireDefault(_tracksTransactions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var EthereumCdpService = (_class = function (_PrivateService) {
  (0, _inherits3.default)(EthereumCdpService, _PrivateService);

  /**
   * @param {string} name
   */
  function EthereumCdpService() {
    var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'cdp';
    (0, _classCallCheck3.default)(this, EthereumCdpService);

    // aliases
    var _this = (0, _possibleConstructorReturn3.default)(this, (EthereumCdpService.__proto__ || (0, _getPrototypeOf2.default)(EthereumCdpService)).call(this, name, ['smartContract', 'token', 'conversion', 'allowance', 'price', 'event']));

    _this.freeEth = _this.freePeth;
    return _this;
  }

  (0, _createClass3.default)(EthereumCdpService, [{
    key: '_smartContract',
    value: function _smartContract() {
      return this.get('smartContract');
    }
  }, {
    key: '_txMgr',
    value: function _txMgr() {
      return this.get('smartContract').get('transactionManager');
    }
  }, {
    key: '_tubContract',
    value: function _tubContract() {
      return this._smartContract().getContractByName(_contracts2.default.SAI_TUB);
    }
  }, {
    key: '_saiProxyTubContract',
    value: function _saiProxyTubContract() {
      return this._smartContract().getContractByName(_contracts2.default.SAI_PROXY);
    }
  }, {
    key: '_web3Service',
    value: function _web3Service() {
      return this._smartContract().get('web3');
    }
  }, {
    key: '_conversionService',
    value: function _conversionService() {
      return this.get('conversion');
    }
  }, {
    key: '_findCdp',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(id) {
        var dsProxyAddress = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
        var info;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(typeof id !== 'number')) {
                  _context.next = 2;
                  break;
                }

                throw new Error('ID must be a number.');

              case 2:
                _context.next = 4;
                return this.getInfo(id);

              case 4:
                info = _context.sent;

                if (!(info.lad.toString() === '0x0000000000000000000000000000000000000000')) {
                  _context.next = 7;
                  break;
                }

                throw new Error("That CDP doesn't exist--try opening a new one.");

              case 7:
                return _context.abrupt('return', dsProxyAddress === null ? new _Cdp2.default(this, id) : new _ProxyCdp2.default(this, dsProxyAddress, id));

              case 8:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function _findCdp(_x2) {
        return _ref.apply(this, arguments);
      }

      return _findCdp;
    }()
  }, {
    key: '_throwIfNotEnoughMkrToWipe',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(cdpId, amountToWipe) {
        var unit = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _Currency.DAI;
        var enoughMkrToWipe;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.enoughMkrToWipe(cdpId, amountToWipe, unit);

              case 2:
                enoughMkrToWipe = _context2.sent;

                if (!(enoughMkrToWipe === false)) {
                  _context2.next = 5;
                  break;
                }

                throw new Error('not enough MKR balance to cover governance fee');

              case 5:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function _throwIfNotEnoughMkrToWipe(_x4, _x5) {
        return _ref2.apply(this, arguments);
      }

      return _throwIfNotEnoughMkrToWipe;
    }()
  }, {
    key: 'enoughMkrToWipe',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(cdpId, amountToWipe) {
        var unit = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _Currency.DAI;

        var dai, MkrToken, ownerAddress, _ref4, _ref5, fee, balance, debt, mkrOwed;

        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                dai = (0, _Currency.getCurrency)(amountToWipe, unit);

                if (!dai.eq(0)) {
                  _context3.next = 3;
                  break;
                }

                return _context3.abrupt('return');

              case 3:
                MkrToken = this.get('token').getToken(_Currency.MKR);
                ownerAddress = this.get('token').get('web3').currentAccount();
                _context3.next = 7;
                return _promise2.default.all([this.getGovernanceFee(cdpId, _Currency.MKR), MkrToken.balanceOf(ownerAddress), this.getDebtValue(cdpId)]);

              case 7:
                _ref4 = _context3.sent;
                _ref5 = (0, _slicedToArray3.default)(_ref4, 3);
                fee = _ref5[0];
                balance = _ref5[1];
                debt = _ref5[2];
                mkrOwed = dai.div(debt).toBigNumber().times(fee.toBigNumber());

                if (!mkrOwed.gt(balance.toBigNumber())) {
                  _context3.next = 15;
                  break;
                }

                return _context3.abrupt('return', false);

              case 15:
                return _context3.abrupt('return', true);

              case 16:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function enoughMkrToWipe(_x7, _x8) {
        return _ref3.apply(this, arguments);
      }

      return enoughMkrToWipe;
    }()
  }, {
    key: 'openCdp',
    value: function openCdp() {
      return new _Cdp2.default(this).transactionObject();
    }
  }, {
    key: 'openProxyCdp',
    value: function openProxyCdp() {
      var dsProxyAddress = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

      return new _ProxyCdp2.default(this, dsProxyAddress).transactionObject();
    }
  }, {
    key: 'getCdp',
    value: function getCdp(id) {
      var dsProxyAddress = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

      return this._findCdp(id, dsProxyAddress);
    }
  }, {
    key: 'shut',
    value: function () {
      var _ref6 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(cdpId) {
        var _this2 = this;

        var debt, hexCdpId;
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.getDebtValue(cdpId, _Currency.DAI);

              case 2:
                debt = _context4.sent;
                _context4.next = 5;
                return this._throwIfNotEnoughMkrToWipe(cdpId, debt);

              case 5:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                return _context4.abrupt('return', _promise2.default.all([this.get('allowance').requireAllowance(_Currency.MKR, this._tubContract().address), this.get('allowance').requireAllowance(_Currency.DAI, this._tubContract().address)]).then(function () {
                  return _this2._tubContract().shut(hexCdpId);
                }));

              case 7:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function shut(_x12) {
        return _ref6.apply(this, arguments);
      }

      return shut;
    }()
  }, {
    key: 'lockEth',
    value: function () {
      var _ref7 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5(cdpId, amount, _ref8) {
        var _ref8$unit = _ref8.unit,
            unit = _ref8$unit === undefined ? _Currency.ETH : _ref8$unit,
            promise = _ref8.promise;
        var convert;
        return _regenerator2.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                convert = this._conversionService().convertEthToWeth(amount, {
                  unit: unit,
                  promise: promise
                });
                _context5.next = 3;
                return this._txMgr().confirm(convert);

              case 3:
                return _context5.abrupt('return', this.lockWeth(cdpId, amount, { promise: promise }));

              case 4:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function lockEth(_x13, _x14, _x15) {
        return _ref7.apply(this, arguments);
      }

      return lockEth;
    }()
  }, {
    key: 'lockWeth',
    value: function () {
      var _ref9 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee6(cdpId, amount, _ref10) {
        var _ref10$unit = _ref10.unit,
            unit = _ref10$unit === undefined ? _Currency.WETH : _ref10$unit,
            promise = _ref10.promise;
        var wethPerPeth, weth;
        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this.get('price').getWethToPethRatio();

              case 2:
                wethPerPeth = _context6.sent;
                weth = (0, _Currency.getCurrency)(amount, unit);
                _context6.next = 6;
                return this._conversionService().convertWethToPeth(weth, {
                  promise: promise
                });

              case 6:
                return _context6.abrupt('return', this.lockPeth(cdpId, weth.div(wethPerPeth), { promise: promise }));

              case 7:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function lockWeth(_x16, _x17, _x18) {
        return _ref9.apply(this, arguments);
      }

      return lockWeth;
    }()
  }, {
    key: 'lockPeth',
    value: function () {
      var _ref11 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee7(cdpId, amount, _ref12) {
        var _ref12$unit = _ref12.unit,
            unit = _ref12$unit === undefined ? _Currency.PETH : _ref12$unit,
            promise = _ref12.promise;
        var hexCdpId, value;
        return _regenerator2.default.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                value = (0, _Currency.getCurrency)(amount, unit).toEthersBigNumber('wei');
                _context7.next = 4;
                return this.get('allowance').requireAllowance(_Currency.PETH, this._tubContract().address, { promise: promise });

              case 4:
                return _context7.abrupt('return', this._tubContract().lock(hexCdpId, value, {
                  promise: promise
                }));

              case 5:
              case 'end':
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function lockPeth(_x19, _x20, _x21) {
        return _ref11.apply(this, arguments);
      }

      return lockPeth;
    }()
  }, {
    key: 'freePeth',
    value: function freePeth(cdpId, amount) {
      var _ref13 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
          _ref13$unit = _ref13.unit,
          unit = _ref13$unit === undefined ? _Currency.PETH : _ref13$unit,
          promise = _ref13.promise;

      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      var value = (0, _Currency.getCurrency)(amount, unit).toEthersBigNumber('wei');
      return this._tubContract().free(hexCdpId, value, { promise: promise });
    }
  }, {
    key: 'drawDai',
    value: function drawDai(cdpId, amount) {
      var _ref14 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
          _ref14$unit = _ref14.unit,
          unit = _ref14$unit === undefined ? _Currency.DAI : _ref14$unit,
          promise = _ref14.promise;

      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      var value = (0, _Currency.getCurrency)(amount, unit).toEthersBigNumber('wei');
      return this._tubContract().draw(hexCdpId, value, { promise: promise });
    }
  }, {
    key: 'wipeDai',
    value: function () {
      var _ref15 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee8(cdpId, amount, _ref16) {
        var _ref16$unit = _ref16.unit,
            unit = _ref16$unit === undefined ? _Currency.DAI : _ref16$unit,
            promise = _ref16.promise;
        var value, hexCdpId;
        return _regenerator2.default.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                value = (0, _Currency.getCurrency)(amount, unit).toEthersBigNumber('wei');
                _context8.next = 3;
                return this._throwIfNotEnoughMkrToWipe(cdpId, amount, unit);

              case 3:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                _context8.next = 6;
                return _promise2.default.all([this.get('allowance').requireAllowance(_Currency.MKR, this._tubContract().address), this.get('allowance').requireAllowance(_Currency.DAI, this._tubContract().address)]);

              case 6:
                return _context8.abrupt('return', this._tubContract().wipe(hexCdpId, value, { promise: promise }));

              case 7:
              case 'end':
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function wipeDai(_x24, _x25, _x26) {
        return _ref15.apply(this, arguments);
      }

      return wipeDai;
    }()
  }, {
    key: 'getInfo',
    value: function getInfo(cdpId) {
      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      return this._tubContract().cups(hexCdpId);
    }
  }, {
    key: 'getCollateralValue',
    value: function () {
      var _ref17 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee9(cdpId) {
        var unit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _Currency.ETH;
        var hexCdpId, pethValue, pethPrice, ethValue, ethPrice, usdValue;
        return _regenerator2.default.wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                _context9.t0 = _Currency.PETH;
                _context9.next = 4;
                return this._tubContract().ink(hexCdpId);

              case 4:
                _context9.t1 = _context9.sent;
                pethValue = _context9.t0.wei.call(_context9.t0, _context9.t1);

                if (!(unit === _Currency.PETH)) {
                  _context9.next = 8;
                  break;
                }

                return _context9.abrupt('return', pethValue);

              case 8:
                _context9.next = 10;
                return this.get('price').getWethToPethRatio();

              case 10:
                pethPrice = _context9.sent;
                ethValue = (0, _Currency.ETH)(pethValue.times(pethPrice));

                if (!(unit === _Currency.ETH)) {
                  _context9.next = 14;
                  break;
                }

                return _context9.abrupt('return', ethValue);

              case 14:
                _context9.next = 16;
                return this.get('price').getEthPrice();

              case 16:
                ethPrice = _context9.sent;
                usdValue = ethValue.times(ethPrice);

                if (!(unit === _Currency.USD)) {
                  _context9.next = 20;
                  break;
                }

                return _context9.abrupt('return', usdValue);

              case 20:
                throw new Error('Don\'t know how to get collateral value in ' + (unit ? unit.symbol : unit));

              case 21:
              case 'end':
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function getCollateralValue(_x27) {
        return _ref17.apply(this, arguments);
      }

      return getCollateralValue;
    }()
  }, {
    key: 'getDebtValue',
    value: function () {
      var _ref18 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee10(cdpId) {
        var unit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _Currency.DAI;
        var hexCdpId, tub, tab, daiDebt, targetPrice;
        return _regenerator2.default.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                // we need to use the Web3.js contract interface to get the return value
                // from the non-constant function `tab`

                tub = this._smartContract().getWeb3ContractByName(_contracts2.default.SAI_TUB);
                _context10.next = 4;
                return new _promise2.default(function (resolve, reject) {
                  return tub.tab.call(hexCdpId, function (err, val) {
                    return err ? reject(err) : resolve(val);
                  });
                });

              case 4:
                tab = _context10.sent;
                daiDebt = _Currency.DAI.wei(tab.toString());
                _context10.t0 = unit;
                _context10.next = _context10.t0 === _Currency.DAI ? 9 : _context10.t0 === _Currency.USD ? 10 : 14;
                break;

              case 9:
                return _context10.abrupt('return', daiDebt);

              case 10:
                _context10.next = 12;
                return this.getTargetPrice();

              case 12:
                targetPrice = _context10.sent;
                return _context10.abrupt('return', daiDebt.times(targetPrice));

              case 14:
              case 'end':
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function getDebtValue(_x29) {
        return _ref18.apply(this, arguments);
      }

      return getDebtValue;
    }()

    //updates compound interest calculations for all CDPs.  Used by tests that depend on a fee

  }, {
    key: '_drip',
    value: function () {
      var _ref19 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee11() {
        return _regenerator2.default.wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                return _context11.abrupt('return', this._tubContract().drip());

              case 1:
              case 'end':
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function _drip() {
        return _ref19.apply(this, arguments);
      }

      return _drip;
    }()
  }, {
    key: 'getGovernanceFee',
    value: function () {
      var _ref20 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee12(cdpId) {
        var unit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _Currency.USD;
        var hexCdpId, tub, rap, usdFee, price;
        return _regenerator2.default.wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                // we need to use the Web3.js contract interface to get the return value
                // from the non-constant function `rap`

                tub = this._smartContract().getWeb3ContractByName(_contracts2.default.SAI_TUB);
                _context12.next = 4;
                return new _promise2.default(function (resolve, reject) {
                  return tub.rap.call(hexCdpId, function (err, val) {
                    return err ? reject(err) : resolve(val);
                  });
                });

              case 4:
                rap = _context12.sent;
                usdFee = _Currency.USD.wei(rap);
                _context12.t0 = unit;
                _context12.next = _context12.t0 === _Currency.USD ? 9 : _context12.t0 === _Currency.MKR ? 10 : 14;
                break;

              case 9:
                return _context12.abrupt('return', usdFee);

              case 10:
                _context12.next = 12;
                return this.get('price').getMkrPrice();

              case 12:
                price = _context12.sent;
                return _context12.abrupt('return', usdFee.div(price));

              case 14:
              case 'end':
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function getGovernanceFee(_x31) {
        return _ref20.apply(this, arguments);
      }

      return getGovernanceFee;
    }()
  }, {
    key: 'getCollateralizationRatio',
    value: function () {
      var _ref21 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee13(cdpId) {
        var usdDebt, _ref22, _ref23, pethPrice, pethCollateral;

        return _regenerator2.default.wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                _context13.next = 2;
                return this.getDebtValue(cdpId, _Currency.USD);

              case 2:
                usdDebt = _context13.sent;

                if (!usdDebt.eq(0)) {
                  _context13.next = 5;
                  break;
                }

                return _context13.abrupt('return', Infinity);

              case 5:
                _context13.next = 7;
                return _promise2.default.all([this.get('price').getPethPrice(), this.getCollateralValue(cdpId, _Currency.PETH)]);

              case 7:
                _ref22 = _context13.sent;
                _ref23 = (0, _slicedToArray3.default)(_ref22, 2);
                pethPrice = _ref23[0];
                pethCollateral = _ref23[1];
                return _context13.abrupt('return', pethCollateral.times(pethPrice).div(usdDebt).toNumber());

              case 12:
              case 'end':
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function getCollateralizationRatio(_x33) {
        return _ref21.apply(this, arguments);
      }

      return getCollateralizationRatio;
    }()
  }, {
    key: 'getLiquidationRatio',
    value: function () {
      var _ref24 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee14() {
        var value;
        return _regenerator2.default.wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                _context14.next = 2;
                return this._tubContract().mat();

              case 2:
                value = _context14.sent;
                return _context14.abrupt('return', new _bignumber2.default(value.toString()).dividedBy(_constants.RAY).toNumber());

              case 4:
              case 'end':
                return _context14.stop();
            }
          }
        }, _callee14, this);
      }));

      function getLiquidationRatio() {
        return _ref24.apply(this, arguments);
      }

      return getLiquidationRatio;
    }()
  }, {
    key: 'getLiquidationPenalty',
    value: function () {
      var _ref25 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee15() {
        var value;
        return _regenerator2.default.wrap(function _callee15$(_context15) {
          while (1) {
            switch (_context15.prev = _context15.next) {
              case 0:
                _context15.next = 2;
                return this._tubContract().axe();

              case 2:
                value = _context15.sent;
                return _context15.abrupt('return', new _bignumber2.default(value.toString()).dividedBy(_constants.RAY).minus(1).toNumber());

              case 4:
              case 'end':
                return _context15.stop();
            }
          }
        }, _callee15, this);
      }));

      function getLiquidationPenalty() {
        return _ref25.apply(this, arguments);
      }

      return getLiquidationPenalty;
    }()
  }, {
    key: 'getTargetPrice',
    value: function () {
      var _ref26 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee16() {
        var vox, par;
        return _regenerator2.default.wrap(function _callee16$(_context16) {
          while (1) {
            switch (_context16.prev = _context16.next) {
              case 0:
                // we need to use the Web3.js contract interface to get the return value
                // from the non-constant function `par()`
                vox = this._smartContract().getWeb3ContractByName(_contracts2.default.SAI_VOX);
                _context16.next = 3;
                return new _promise2.default(function (resolve, reject) {
                  return vox.par.call(function (err, val) {
                    return err ? reject(err) : resolve(val);
                  });
                });

              case 3:
                par = _context16.sent;
                return _context16.abrupt('return', _Currency.USD_DAI.ray(par));

              case 5:
              case 'end':
                return _context16.stop();
            }
          }
        }, _callee16, this);
      }));

      function getTargetPrice() {
        return _ref26.apply(this, arguments);
      }

      return getTargetPrice;
    }()
  }, {
    key: 'getLiquidationPrice',
    value: function () {
      var _ref27 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee17(cdpId) {
        var _ref28, _ref29, debt, liqRatio, collateral;

        return _regenerator2.default.wrap(function _callee17$(_context17) {
          while (1) {
            switch (_context17.prev = _context17.next) {
              case 0:
                _context17.next = 2;
                return _promise2.default.all([this.getDebtValue(cdpId, _Currency.USD), this.getLiquidationRatio(), this.getCollateralValue(cdpId)]);

              case 2:
                _ref28 = _context17.sent;
                _ref29 = (0, _slicedToArray3.default)(_ref28, 3);
                debt = _ref29[0];
                liqRatio = _ref29[1];
                collateral = _ref29[2];

                if (!collateral.eq(0)) {
                  _context17.next = 9;
                  break;
                }

                return _context17.abrupt('return', (0, _Currency.USD_ETH)(Infinity));

              case 9:
                return _context17.abrupt('return', debt.times(liqRatio).div(collateral));

              case 10:
              case 'end':
                return _context17.stop();
            }
          }
        }, _callee17, this);
      }));

      function getLiquidationPrice(_x34) {
        return _ref27.apply(this, arguments);
      }

      return getLiquidationPrice;
    }()
  }, {
    key: 'isSafe',
    value: function () {
      var _ref30 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee18(cdpId) {
        var _ref31, _ref32, liqPrice, ethPrice;

        return _regenerator2.default.wrap(function _callee18$(_context18) {
          while (1) {
            switch (_context18.prev = _context18.next) {
              case 0:
                _context18.next = 2;
                return _promise2.default.all([this.getLiquidationPrice(cdpId), this.get('price').getEthPrice()]);

              case 2:
                _ref31 = _context18.sent;
                _ref32 = (0, _slicedToArray3.default)(_ref31, 2);
                liqPrice = _ref32[0];
                ethPrice = _ref32[1];
                return _context18.abrupt('return', (0, _Currency.USD_ETH)(ethPrice).gte(liqPrice));

              case 7:
              case 'end':
                return _context18.stop();
            }
          }
        }, _callee18, this);
      }));

      function isSafe(_x35) {
        return _ref30.apply(this, arguments);
      }

      return isSafe;
    }()
  }, {
    key: 'getAnnualGovernanceFee',
    value: function getAnnualGovernanceFee() {
      return this._tubContract().fee().then(function (bn) {
        var fee = new _bignumber2.default(bn.toString()).dividedBy(_constants.RAY);
        var secondsPerYear = 60 * 60 * 24 * 365;
        _bignumber2.default.config({ POW_PRECISION: 100 });
        return fee.pow(secondsPerYear).minus(1).toNumber();
      });
    }
  }, {
    key: 'getSystemCollateralization',
    value: function () {
      var _ref33 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee19() {
        var dai, _ref34, _ref35, totalWethLocked, wethPrice, daiSupply, targetPrice, totalCollateralValue, systemDaiDebt;

        return _regenerator2.default.wrap(function _callee19$(_context19) {
          while (1) {
            switch (_context19.prev = _context19.next) {
              case 0:
                dai = this.get('token').getToken(_Currency.DAI);
                _context19.next = 3;
                return _promise2.default.all([this._tubContract().pie(), this.get('price').getEthPrice(), dai.totalSupply(), this.getTargetPrice()]);

              case 3:
                _ref34 = _context19.sent;
                _ref35 = (0, _slicedToArray3.default)(_ref34, 4);
                totalWethLocked = _ref35[0];
                wethPrice = _ref35[1];
                daiSupply = _ref35[2];
                targetPrice = _ref35[3];
                totalCollateralValue = new _bignumber2.default(totalWethLocked).div(_constants.WAD).times(wethPrice.toBigNumber());
                systemDaiDebt = daiSupply.times(targetPrice);
                return _context19.abrupt('return', totalCollateralValue.div(systemDaiDebt.toBigNumber()).toNumber());

              case 12:
              case 'end':
                return _context19.stop();
            }
          }
        }, _callee19, this);
      }));

      function getSystemCollateralization() {
        return _ref33.apply(this, arguments);
      }

      return getSystemCollateralization;
    }()
  }, {
    key: 'getWethToPethRatio',
    value: function () {
      var _ref36 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee20() {
        var value;
        return _regenerator2.default.wrap(function _callee20$(_context20) {
          while (1) {
            switch (_context20.prev = _context20.next) {
              case 0:
                _context20.next = 2;
                return this._tubContract().per();

              case 2:
                value = _context20.sent;
                return _context20.abrupt('return', new _bignumber2.default(value.toString()).dividedBy(_constants.RAY).toNumber());

              case 4:
              case 'end':
                return _context20.stop();
            }
          }
        }, _callee20, this);
      }));

      function getWethToPethRatio() {
        return _ref36.apply(this, arguments);
      }

      return getWethToPethRatio;
    }()
  }, {
    key: 'give',
    value: function give(cdpId, newAddress) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      return this._tubContract().give(hexCdpId, newAddress, options);
    }
  }, {
    key: 'bite',
    value: function bite(cdpId) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      return this._tubContract().bite(hexCdpId, options);
    }
  }, {
    key: 'freeEthProxy',
    value: function freeEthProxy(dsProxyAddress, cdpId, amount) {
      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      var value = (0, _Currency.getCurrency)(amount, _Currency.ETH).toEthersBigNumber('wei');

      return this._saiProxyTubContract().free(this._tubContract().address, hexCdpId, value, {
        dsProxyAddress: dsProxyAddress,
        metadata: {
          action: {
            name: 'free',
            id: cdpId,
            amount: (0, _Currency.getCurrency)(amount, _Currency.ETH),
            proxy: true
          }
        }
      });
    }
  }, {
    key: 'lockEthProxy',
    value: function lockEthProxy(dsProxyAddress, cdpId, amount) {
      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      var value = (0, _Currency.getCurrency)(amount, _Currency.ETH).toEthersBigNumber('wei');

      return this._saiProxyTubContract().lock(this._tubContract().address, hexCdpId, {
        dsProxyAddress: dsProxyAddress,
        value: value,
        metadata: {
          action: {
            name: 'lock',
            id: cdpId,
            amount: (0, _Currency.getCurrency)(amount, _Currency.ETH),
            proxy: true
          }
        }
      });
    }
  }, {
    key: 'drawDaiProxy',
    value: function drawDaiProxy(dsProxyAddress, cdpId, amount) {
      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
      var value = (0, _Currency.getCurrency)(amount, _Currency.DAI).toEthersBigNumber('wei');

      return this._saiProxyTubContract().draw(this._tubContract().address, hexCdpId, value, {
        dsProxyAddress: dsProxyAddress,
        metadata: {
          action: {
            name: 'draw',
            id: cdpId,
            amount: (0, _Currency.getCurrency)(amount, _Currency.DAI),
            proxy: true
          }
        }
      });
    }
  }, {
    key: 'giveProxy',
    value: function giveProxy(dsProxyAddress, cdpId, newAddress) {
      var hexCdpId = (0, _conversion.numberToBytes32)(cdpId);

      return this._saiProxyTubContract().give(this._tubContract().address, hexCdpId, newAddress, {
        dsProxyAddress: dsProxyAddress,
        metadata: {
          action: {
            name: 'give',
            id: cdpId,
            to: newAddress,
            proxy: true
          }
        }
      });
    }
  }, {
    key: 'wipeDaiProxy',
    value: function () {
      var _ref37 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee21(dsProxyAddress, cdpId, amount, _ref38) {
        var useOtc = _ref38.useOtc,
            promise = _ref38.promise;
        var hexCdpId, value, options;
        return _regenerator2.default.wrap(function _callee21$(_context21) {
          while (1) {
            switch (_context21.prev = _context21.next) {
              case 0:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                value = (0, _Currency.getCurrency)(amount, _Currency.DAI).toEthersBigNumber('wei');
                _context21.next = 4;
                return this.get('allowance').requireAllowance(_Currency.DAI, dsProxyAddress, {
                  promise: promise
                });

              case 4:
                if (useOtc) {
                  _context21.next = 9;
                  break;
                }

                _context21.next = 7;
                return this._throwIfNotEnoughMkrToWipe(cdpId, amount, _Currency.DAI);

              case 7:
                _context21.next = 9;
                return this.get('allowance').requireAllowance(_Currency.MKR, dsProxyAddress, {
                  promise: promise
                });

              case 9:
                options = {
                  dsProxyAddress: dsProxyAddress,
                  metadata: {
                    action: {
                      name: 'wipe',
                      id: cdpId,
                      amount: (0, _Currency.getCurrency)(amount, _Currency.DAI),
                      otc: useOtc,
                      proxy: true
                    }
                  },
                  promise: promise
                };

                // If using OTC to buy MKR to pay fee, pass OTC address to SaiProxy wipe()
                // method

                return _context21.abrupt('return', useOtc ? this._saiProxyTubContract()['wipe(address,bytes32,uint256,address)'](this._tubContract().address, hexCdpId, value, this._smartContract().getContractAddressByName(_contracts2.default.MAKER_OTC), options) : this._saiProxyTubContract()['wipe(address,bytes32,uint256)'](this._tubContract().address, hexCdpId, value, options));

              case 11:
              case 'end':
                return _context21.stop();
            }
          }
        }, _callee21, this);
      }));

      function wipeDaiProxy(_x38, _x39, _x40, _x41) {
        return _ref37.apply(this, arguments);
      }

      return wipeDaiProxy;
    }()
  }, {
    key: 'shutProxy',
    value: function () {
      var _ref39 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee22(dsProxyAddress, cdpId, _ref40) {
        var useOtc = _ref40.useOtc,
            promise = _ref40.promise;
        var hexCdpId, debt, options;
        return _regenerator2.default.wrap(function _callee22$(_context22) {
          while (1) {
            switch (_context22.prev = _context22.next) {
              case 0:
                hexCdpId = (0, _conversion.numberToBytes32)(cdpId);
                _context22.next = 3;
                return this.get('allowance').requireAllowance(_Currency.DAI, dsProxyAddress, {
                  promise: promise
                });

              case 3:
                if (useOtc) {
                  _context22.next = 11;
                  break;
                }

                _context22.next = 6;
                return this.getDebtValue(cdpId, _Currency.DAI);

              case 6:
                debt = _context22.sent;
                _context22.next = 9;
                return this._throwIfNotEnoughMkrToWipe(cdpId, debt);

              case 9:
                _context22.next = 11;
                return this.get('allowance').requireAllowance(_Currency.MKR, dsProxyAddress, {
                  promise: promise
                });

              case 11:
                options = {
                  dsProxyAddress: dsProxyAddress,
                  metadata: {
                    action: {
                      name: 'close',
                      id: cdpId,
                      otc: useOtc,
                      proxy: true
                    }
                  },
                  promise: promise
                };

                // If using OTC to buy MKR to pay fee, pass OTC address to SaiProxy shut()
                // method

                return _context22.abrupt('return', useOtc ? this._saiProxyTubContract()['shut(address,bytes32,address)'](this._tubContract().address, hexCdpId, this._smartContract().getContractAddressByName(_contracts2.default.MAKER_OTC), options) : this._saiProxyTubContract()['shut(address,bytes32)'](this._tubContract().address, hexCdpId, options));

              case 13:
              case 'end':
                return _context22.stop();
            }
          }
        }, _callee22, this);
      }));

      function shutProxy(_x42, _x43, _x44) {
        return _ref39.apply(this, arguments);
      }

      return shutProxy;
    }()
  }]);
  return EthereumCdpService;
}(_PrivateService3.default), (_applyDecoratedDescriptor(_class.prototype, 'lockEth', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'lockEth'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'lockWeth', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'lockWeth'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'lockPeth', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'lockPeth'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'wipeDai', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'wipeDai'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'wipeDaiProxy', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'wipeDaiProxy'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'shutProxy', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'shutProxy'), _class.prototype)), _class);
exports.default = EthereumCdpService;
