// @makerdao/dai/contracts/networks.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TESTNET_ID = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

exports.contractInfo = contractInfo;

var _tokens = require('./tokens');

var _tokens2 = _interopRequireDefault(_tokens);

var _contracts = require('./contracts');

var _contracts2 = _interopRequireDefault(_contracts);

var _abis = require('./abis');

var abis = _interopRequireWildcard(_abis);

var _kovan = require('./addresses/kovan.json');

var _kovan2 = _interopRequireDefault(_kovan);

var _mainnet = require('./addresses/mainnet.json');

var _mainnet2 = _interopRequireDefault(_mainnet);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function contractInfo(network) {
  var _ref;

  var addresses = void 0;
  switch (network) {
    case 'mainnet':
      addresses = _mainnet2.default;
      break;
    case 'kovan':
      addresses = _kovan2.default;
      break;
    case 'test':
      // we "require just in time" here because a non-developer user will not
      // have a testnet addresses file
      addresses = require('./addresses/testnet.json');
      break;
  }

  return _ref = {}, (0, _defineProperty3.default)(_ref, _tokens2.default.DAI, [{
    version: 1,
    address: addresses.SAI,
    abi: abis.general.erc20,
    decimals: 18
  }]), (0, _defineProperty3.default)(_ref, _tokens2.default.WETH, [{
    version: 1,
    address: addresses.GEM,
    abi: abis.dappHub.dsEthToken,
    decimals: 18
  }]), (0, _defineProperty3.default)(_ref, _tokens2.default.PETH, [{
    version: 1,
    address: addresses.SKR,
    abi: abis.general.erc20,
    decimals: 18
  }]), (0, _defineProperty3.default)(_ref, _tokens2.default.MKR, [{
    version: 1,
    address: addresses.GOV_OLD || '0x0000000000000000000000000000000000000001',
    abi: abis.general.erc20,
    decimals: 18
  }, {
    version: 2,
    address: addresses.GOV,
    abi: abis.general.erc20,
    decimals: 18
  }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_TOP, [{ version: 1, address: addresses.TOP, abi: abis.daiV1.saiTop }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_PIP, [{ version: 1, address: addresses.PIP, abi: abis.dappHub.dsValue }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_PEP, [{ version: 1, address: addresses.PEP, abi: abis.dappHub.dsValue }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_PIT, [{ version: 1, address: addresses.PIT, abi: abis.daiV1.pit }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_SIN, [{ version: 1, address: addresses.SIN, abi: abis.general.erc20 }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_DAD, [{ version: 1, address: addresses.DAD, abi: abis.dappHub.dsGuard }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_MOM, [{ version: 1, address: addresses.MOM, abi: abis.daiV1.mom }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_VOX, [{ version: 1, address: addresses.VOX, abi: abis.daiV1.vox }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_TAP, [{ version: 1, address: addresses.TAP, abi: abis.daiV1.tap }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_TUB, [{ version: 1, address: addresses.TUB, abi: abis.daiV1.tub }]), (0, _defineProperty3.default)(_ref, _contracts2.default.MAKER_OTC, [{
    version: 1,
    address: addresses.MAKER_OTC,
    abi: abis.exchangesV1.makerOtc
  }]), (0, _defineProperty3.default)(_ref, _contracts2.default.SAI_PROXY, [{
    version: 1,
    address: addresses.SAI_PROXY,
    abi: abis.proxies.saiProxy
  }]), (0, _defineProperty3.default)(_ref, _contracts2.default.DS_PROXY_FACTORY, [{
    version: 1,
    address: addresses.DS_PROXY_FACTORY,
    abi: abis.proxies.dsProxyFactory
  }]), (0, _defineProperty3.default)(_ref, _contracts2.default.DS_PROXY, [{
    version: 1,
    abi: abis.dappHub.dsProxy
  }]), (0, _defineProperty3.default)(_ref, _contracts2.default.PROXY_REGISTRY, [{
    version: 1,
    address: addresses.PROXY_REGISTRY,
    abi: abis.proxies.proxyRegistry
  }]), _ref;
}

var TESTNET_ID = exports.TESTNET_ID = 999;

exports.default = [{ name: 'mainnet', networkId: 1, contracts: contractInfo('mainnet') }, { name: 'morden', networkId: 2, contracts: null }, { name: 'ropsten', networkId: 3, contracts: null }, { name: 'rinkeby', networkId: 4, contracts: null }, { name: 'kovan', networkId: 42, contracts: contractInfo('kovan') }, { name: 'test', networkId: TESTNET_ID, contracts: contractInfo('test') }];
