// @makerdao/dai/src/exchanges/oasis/OasisOrder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OasisSellOrder = exports.OasisBuyOrder = exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _ethers = require('ethers');

var _Currency = require('../../eth/Currency');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OasisOrder = function () {
  function OasisOrder() {
    (0, _classCallCheck3.default)(this, OasisOrder);
  }

  (0, _createClass3.default)(OasisOrder, [{
    key: 'fillAmount',
    value: function fillAmount() {
      return this._fillAmount;
    }
  }, {
    key: 'fees',
    value: function fees() {
      return this._txMgr.getTransaction(this.promise).fees();
    }
  }, {
    key: 'created',
    value: function created() {
      return this._txMgr.getTransaction(this.promise).timestamp();
    }
  }, {
    key: 'transact',
    value: function transact(contract, method, args, transactionManager) {
      var _this = this;

      this._contract = contract;
      this._txMgr = transactionManager;
      var promise = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
        var txo;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return 0;

              case 2:
                _context.next = 4;
                return contract[method].apply(contract, [].concat((0, _toConsumableArray3.default)(args), [{ promise: promise }]));

              case 4:
                txo = _context.sent;

                _this._parseLogs(txo.receipt.logs);
                return _context.abrupt('return', _this);

              case 7:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, _this);
      }))();
      this.promise = promise;
      return promise;
    }
  }, {
    key: '_parseLogs',
    value: function _parseLogs(logs) {
      var _this2 = this;

      var LogTrade = this._contract.interface.events.LogTrade;

      // TODO convert string to hex without web3

      var topic = _ethers.utils.keccak256(this._txMgr.get('web3')._web3.toHex(LogTrade.signature));

      var receiptEvents = logs.filter(function (e) {
        return e.topics[0].toLowerCase() === topic.toLowerCase() && e.address.toLowerCase() === _this2._contract.address.toLowerCase();
      });

      var total = receiptEvents.reduce(function (acc, event) {
        var parsedLog = LogTrade.parse(event.data);
        return acc.add(parsedLog[_this2._logKey]);
      }, _ethers.utils.bigNumberify('0'));
      this._fillAmount = this._unit.wei(total.toString());
    }
  }]);
  return OasisOrder;
}();

exports.default = OasisOrder;

var OasisBuyOrder = exports.OasisBuyOrder = function (_OasisOrder) {
  (0, _inherits3.default)(OasisBuyOrder, _OasisOrder);

  function OasisBuyOrder() {
    (0, _classCallCheck3.default)(this, OasisBuyOrder);

    var _this3 = (0, _possibleConstructorReturn3.default)(this, (OasisBuyOrder.__proto__ || (0, _getPrototypeOf2.default)(OasisBuyOrder)).call(this));

    _this3._logKey = 'buy_amt';
    _this3._unit = _Currency.DAI;
    return _this3;
  }

  (0, _createClass3.default)(OasisBuyOrder, null, [{
    key: 'build',
    value: function build(contract, method, args, transactionManager) {
      var order = new OasisBuyOrder();
      order.transact(contract, method, args, transactionManager);
      return order.promise;
    }
  }]);
  return OasisBuyOrder;
}(OasisOrder);

var OasisSellOrder = exports.OasisSellOrder = function (_OasisOrder2) {
  (0, _inherits3.default)(OasisSellOrder, _OasisOrder2);

  function OasisSellOrder(currency) {
    (0, _classCallCheck3.default)(this, OasisSellOrder);

    var _this4 = (0, _possibleConstructorReturn3.default)(this, (OasisSellOrder.__proto__ || (0, _getPrototypeOf2.default)(OasisSellOrder)).call(this));

    _this4._logKey = 'pay_amt';
    _this4._unit = currency;
    return _this4;
  }

  (0, _createClass3.default)(OasisSellOrder, null, [{
    key: 'build',
    value: function build(contract, method, args, transactionManager, currency) {
      var order = new OasisSellOrder(currency);
      order.transact(contract, method, args, transactionManager);
      return order.promise;
    }
  }]);
  return OasisSellOrder;
}(OasisOrder);
