// @makerdao/dai/src/eth/tokens/EtherToken.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _maxSafeInteger = require('babel-runtime/core-js/number/max-safe-integer');

var _maxSafeInteger2 = _interopRequireDefault(_maxSafeInteger);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _desc, _value, _class;

var _ethers = require('ethers');

var _Currency = require('../Currency');

var _tracksTransactions = require('../../utils/tracksTransactions');

var _tracksTransactions2 = _interopRequireDefault(_tracksTransactions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var EtherToken = (_class = function () {
  function EtherToken(web3Service, gasEstimatorService, transactionManager) {
    (0, _classCallCheck3.default)(this, EtherToken);

    this._web3 = web3Service;
    this._gasEstimator = gasEstimatorService;
    this._transactionManager = transactionManager;
  }

  // eslint-disable-next-line


  (0, _createClass3.default)(EtherToken, [{
    key: 'allowance',
    value: function allowance(tokenOwner, spender) {
      return _promise2.default.resolve(_maxSafeInteger2.default);
    }
  }, {
    key: 'balanceOf',
    value: function balanceOf(owner) {
      return this._web3.ethersProvider().getBalance(owner).then(function (b) {
        return _ethers.utils.formatEther(b);
      });
    }

    // eslint-disable-next-line

  }, {
    key: 'approve',
    value: function approve(spender, value) {
      return _promise2.default.resolve(true);
    }

    // eslint-disable-next-line

  }, {
    key: 'approveUnlimited',
    value: function approveUnlimited(spender) {
      return _promise2.default.resolve(true);
    }
  }, {
    key: 'transfer',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(toAddress, amount, options) {
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                return _context.abrupt('return', this.transferFrom(this._web3.currentAccount(), toAddress, amount, options));

              case 1:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function transfer(_x, _x2, _x3) {
        return _ref.apply(this, arguments);
      }

      return transfer;
    }()
  }, {
    key: 'transferFrom',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(fromAddress, toAddress, amount, _ref3) {
        var _ref3$unit = _ref3.unit,
            unit = _ref3$unit === undefined ? _Currency.ETH : _ref3$unit,
            promise = _ref3.promise;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                return _context2.abrupt('return', this._transactionManager.sendTransaction({
                  from: fromAddress,
                  to: toAddress,
                  value: (0, _Currency.getCurrency)(amount, unit).toEthersBigNumber('wei').toString()
                }, { promise: promise }));

              case 1:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function transferFrom(_x4, _x5, _x6, _x7) {
        return _ref2.apply(this, arguments);
      }

      return transferFrom;
    }()
  }]);
  return EtherToken;
}(), (_applyDecoratedDescriptor(_class.prototype, 'transfer', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'transfer'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'transferFrom', [_tracksTransactions2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'transferFrom'), _class.prototype)), _class);
exports.default = EtherToken;
